### DISCUSSION TITLE
How to write to files that require super user rights?

### DISCUSSION THREAD

**Discussion by kirusfg** [helix-editor/helix#4251] [Q&A] [answered]
When editing a file that does not belong to me, I do `sudo helix file` so that I have rights to write to it, but the config loaded in helix is no longer mine. Is there a way to "sudo-write" from a user instance of `helix`?

**Comment by the-mikedavis** [comment, marked as answer]
[`sudoedit(8)`](https://man7.org/linux/man-pages/man8/sudoedit.8.html) is a good tool for this: instead of running `sudo helix` which runs helix with root permissions, it creates a temp file, runs `helix` as you, and then moves the temp file to the location of the real file when done editing.

**Reply by kirusfg** [reply]
Oh, yes, I know of that command. It would be nice if helix could do the same.

**Reply by herlev** [reply]
I often only realise I don't have permissions to write the file after making my changes. I then copy the entire file to clipboard, quit, edit with sudo, paste clipboard contents and save. 

The command `:SudoWrite` from [tpope/vim-eunuch](https://github.com/tpope/vim-eunuch) would greatly improve improve my workflow if implemented for Helix, either as a plugin or natively. After getting a permission denied error from `:w`, I will just type `:SudoWrite`, type in my password when prompted and the file is written.

**Reply by David-Else** [reply]
Maybe a better solution would be to make Helix aware if it was called as `SUDO_EDITOR` and simply display in the status line 'You don't have permission to modify this file' if that were the case?

**Reply by sudormrfbin** [reply]
Upon loading a new document we could check whether the current user has write permissions to the file and disable editing. Vim for example shows `Warning: Editing a readonly file` when attempting to change it (but still allows editing the text in the buffer).

**Reply by the-mikedavis** [reply]
A read-only indicator in the statusline would be a nice touch I think (https://github.com/helix-editor/helix/issues/2627). Currently we only show an error when attempting to write to a read-only file.

**Reply by kirusfg** [reply]
I would personally prefer something like `:sudo-write` since that is what I use in vim and kakoune. But in those editors, these are plugins, not the editors' core functionality.

**Reply by yonatan8070** [reply]
I know that VS Code has a similar feature, if you try to save a file without write permissions, it will offer to save as root (using pkexec), save to a different location, or discard. I think this could be a really useful feature.

**Comment by David-Else** [comment]
To make helix your default sudo editor edit your `.bash_profile` and add:

```
export SUDO_EDITOR="path/to/hx"
```
Then use `sudoedit`.

**Reply by pwnspirit** [reply]
same as
```bash
export SUDO_EDITOR=$(which hx) 
``` 
and ``sudoedit``

**Comment by pathwave** [comment]
When you run helix as sudo helix look for config in root/.config/helix/.. path. You can supply your own config path by doing hx --config /my/path/config.toml or symlink config to root profile folder.

Another option is to do what installers do and make a bash script in for example /usr/bin/ named hx, with content like this:
#!/bin/sh
HELIX_RUNTIME=/path/to/helix/runtime exec /path/to/real/helix/hx --config /path/to/helix/config.toml "$@"

**Comment by fraunos** [comment]
micro editor asks for root password when the user is not permitted to write the file. This is really convenient.

**Reply by webdev23** [reply]
It also watch and warn if the file has changed, it's critical in this context.

**Reply by zhexyez** [reply]
> micro editor asks for root password when the user is not permitted to write the file. This is really convenient.

Similar behavior would be appreciated, as `SUDO_EDITOR` does not warn that you want to write into the file that needs root permissions.

**Comment by uros-5** [comment]
If changing `EDITOR` and `SUDO_EDITOR` don't work, run this command:
```bash
sudo update-alternatives --install /usr/bin/editor editor /home/username/.cargo/bin/hx 50
```
then select helix:
```bash
sudo update-alternatives --config editor
```

**Comment by uros-5** [comment]
Run which hx to see where is helix binary. And just replace output from
that command into first one.

**Reply by nowhereman1999** [reply]
Sure, appreciate your prompt response. I've actually discovered that the helix binary is located at `/usr/bin/hx`, confirmed by using the `which hx` command. However, even after running the command with the updated path, the configuration didn't load when I tried opening it with `sudo hx`, as I explained in this post: [link](https://github.com/helix-editor/helix/discussions/7163#discussioncomment-6639581).

**Comment by webdev23** [comment]
Good care has to be taken on all this little details, as it breaks the user experience real bad.

Our keybindings goes away and we get stuck on the pink fantasy theme. It's unusable, we close it.

**Comment by HernandoR** [comment]
in vim, we could use `:w !sudo tee % ` when we find out the requirement for sudo permission after editing. Is there a similar tool in hx?

**Reply by yonatan8070** [reply]
I think you can use `%:pipe-to sudo tee %`, haven't tested it but it should work

**Reply by jobukkit** [reply]
> I think you can use `%:pipe-to sudo tee %`, haven't tested it but it should work

Didn't work for me, there was no error but the changes weren't written. Don't know if I did something wrong.

**Reply by curiositry** [reply]
@yonatan8070 @jobukkit  `%:pipe-to sudo tee %` seems to write the current file to a file named `%`. What I used was `%:pipe-to sudo tee <C-r>%` (ctrl+r % outputs the current file path from the % register when in command mode.)

**Reply by nothingnesses** [reply]
For posterity, the solution by @curiositry works, but be very careful that you do this properly, in case you end up doing what I did: I forgot to select all the text using "%" before using `pipe-to` and didn't realise until after I had saved my NixOS configuration and closed the editor. As you can guess by now, this ended up writing an empty file. Unfortunately, I had no recent backups, the configs weren't in VCS, ext4 doesn't have a snapshot/rollback feature and I had recently GC'ed "/nix/store". I spent almost a whole day to recover the file contents, which thankfully I was able to do. I've since learnt my lesson and am now keeping my config in VCS + I keep regular backups.

That being said, this operation (saving a file as superuser) really should be much easier to do, perhaps added as a built-in command like `pipe-to`, instead of having to rely solely on the user to correctly invoke a sequence of commands, which can easily be mistyped. `micro` handles this very cleanly: when trying to save a file without superuser access, the program will prompt the user for a password to escalate their privilege. Despite how much I like the rest of the editor, I don't understand why `hx` still doesn't do something similar after all this time.

**Comment by htqx** [comment]
....Not as good as micro
